### DISCUSSION TITLE
[Question]: Is there interest in accepting Typescript backend PRs ?

### DISCUSSION THREAD

**Discussion by publicvirtualvoid** [danny-avila/LibreChat#1224] [Q&A] [answered]
### What is your question?

I'm interested in contributing, and I believe typing will help make the code more accessible. Is there an appetite for making use of TS on the backend?

### More Details

Having read the considerations in the Contributing Guidelines, I personally don't believe there is an advantage to making many changes in one go. The first change  would be to configure TS to include `.js` files and integrate the build step into the existing commands. From there files can be gradually transitioned as needed. 

Not looking get overzealous with types but signatures on functions would be a great to have.

Any problems in my reasoning here?

### What is the main subject of your question?

Other

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by danny-avila** [comment, marked as answer]
Unfortunately this could be too much of an undertaking for not a lot of gain, especially since some of the modules are built with CJS specific code, and figuring them out in ESM could be involved or rather not worth the effort.

> signatures on functions would be a great to have.

I 100% agree with you here; though Typescript safety is nice (I like it a lot on the frontend), I've begun to write out more JSDoc and make use of typedefs to benefit from this. This even allows me to re-use some of the types from the frontend. See `api/typedefs.js`

To this end, I would very much appreciate help in adding more JSDoc or improving what's already there

Some resources that have bolstered my opinion against an effort to convert the project to ESM/TS:
- [CommonJS is not going away](https://bun.sh/blog/commonjs-is-not-going-away)
- [TypeScript is ‘not worth it’ for developing libraries, says Svelte author, as team switches to JavaScript and JSDoc](https://news.ycombinator.com/item?id=35932617)
- [Rich Harris, creator of Svelte, on switch](https://news.ycombinator.com/item?id=35892250)

**Comment by publicvirtualvoid** [comment]
Thanks, that seems a reasonable compromise for the time being.
